File: static/game/GameClient/Objects/Game/Game.js
import {EventManager} from "../../Components/EventManager/EventManager.js";
import {GameField} from "../GameField/GameField.js";
import {ServerMessagesQueue} from "../../Components/ServerMessagesQueue/ServerMessagesQueue.js";
import {UI} from "../UI/UI.js";
import {Notificator} from "../Notificator/Notificator.js";

import {Figures} from "../Figures/Figures.js";
import {Constants} from "../Constants.js";
import {GlobalGameStateController} from "../GlobalGameStateController/GlobalGameStateController.js";

export class Game{
    constructor(context, websocket, window_rect){
        this.context = context;
        this.ws = websocket;
        this.window_rect = window_rect;
        this.event_manager = new EventManager();
        this.game_field = new GameField("/static/game/Images/bg.png");
        this.ui = new UI();
        this.closed = false;
        this.notificator = new Notificator();
    }

    handleEvent(event_name, event_body){
        if(event_body.code == "Space")
        {
            this.game_field.end_turn();
            this.sendMessagesToServer();
        }
        if(event_body.key == "Backspace")
        {
            this.ws.send(JSON.stringify({
                'type': 'close'
            }));
            GlobalGameStateController.changeGameState("menu");
        }
    }

    sendMessagesToServer() {
        this.ws.send(JSON.stringify({
            'type': 'game',
            'user': Constants.game_master,
            'actions': ServerMessagesQueue.to_send
        }));
        ServerMessagesQueue.to_send = [];
    }

    handleServerRequest(data){
        this.ui.infobar.turn = data['turn'];

        if(data['user'] != Constants.game_master){
            for(let action of data['actions']){
                if (action['master'] != Constants.game_master) {
                    setTimeout(() => {
                        let from_x = action['from_x'];
                        let from_y = action['from_y'];
                        let to_x = action['to_x'];
                        let to_y = action['to_y'];
                        from_x = data['left'] != Constants.game_master ? Constants.x_field_block_amount - from_x - 1 : from_x;
                        to_x = data['left'] != Constants.game_master ? Constants.x_field_block_amount - to_x - 1 : to_x;
                        switch (action['type']) {
                            case 'attack': {
                                this.game_field.attackFigure(from_x, from_y, to_x, to_y);
                            }
                                break;
                            case 'move': {
                                this.game_field.moveObject(from_x, from_y, to_x, to_y);
                            }
                                break;
                        }
                    }, 0);
                }
            }
        }

        setTimeout(()=>{
            this.game_field.setTurn(data['turn']);
        }, 0);
    }

    renderAll(){
        this.game_field.render(this.context);
        this.ui.render(this.context);
        this.notificator.render(this.context);
    }

    init(server_data){
        this.ui.infobar.turn = server_data['turn'];

        setTimeout(()=> {
            for (let figure of server_data['field']) {
                let x = figure['x'];
                if (figure['master'] == Constants.game_master) {
                    x = x > Math.floor(Constants.x_field_block_amount / 2) ? Constants.x_field_block_amount - x - 1 : x;
                } else {
                    x = x < Math.round(Constants.x_field_block_amount / 2) ? Constants.x_field_block_amount - x - 1 : x;
                }
                let object = new (Figures.get_figure(figure['name']))(figure['master'] != Constants.game_master);
                object.setProps(figure['props']);
                this.game_field.setObject(object, x, figure['y']);
            }
        }, 0);
        setTimeout(()=>{
            this.game_field.setTurn(server_data['turn']);
        }, 0);
        this.start();
    }

    start(){
        this.event_manager.subscribe('keydown', this);
        this.event_manager.subscribe("mousedown", this.game_field);
        this.event_manager.subscribe("mouseup", this.game_field);
        this.event_manager.subscribe("keydown", this.game_field);
        this.event_manager.subscribe("mousedown", this.ui.infobar.btn_back);
        this.event_manager.subscribe("mouseup", this.ui.infobar.btn_back);
        this.event_manager.subscribe("mousemove", this.ui.infobar.btn_back);

        for(let i = 0; i < this.game_field.field_matrix.length; ++i){
            for(let j = 0; j < this.game_field.field_matrix[0].length; ++j){
                this.event_manager.subscribe("mousemove", this.game_field.field_matrix[i][j]);
            }
        }

        //starting game cycle
        let game_cycle = ()=>{
            this.renderAll();
            if(!this.closed) {
                requestAnimationFrame(game_cycle);
            }else{
                this.context.clearRect(this.window_rect.x, this.window_rect.y, this.window_rect.width, this.window_rect.height);
            }
        }
        requestAnimationFrame(game_cycle);
    }

    end(){
        this.closed = true;
        this.ws.send(JSON.stringify({
                'type': 'close'
            }));
    }
}

File: game/GameServer/Game.py
import random as rand
from game.GameServer.Objects.GameField.GameField import GameField
from game.GameServer.Objects.Figures.Figures import Figures
from game.GameServer.Constants import Constants
from game.GameServer.Objects.GameObject.GameObject import GameObject
from game.GameServer.Objects.Figure.Figure import Figure
from ..models import Figure as fg


class Game:
    def __init__(self, callback):
        self.turn = rand.randint(0, 1)
        self.game_field = GameField()
        self.callback = callback
        self.left = None
        self.right = None
        self.left_force = 0
        self.right_force = 0
        self.won_prize = 0
        self.winner = None

    def init(self, data, figures):
        if self.left is None:
            self.left = data['user']
            for figure in figures:                
                self.game_field.setFigure(Figures.get_figure(name=figure.figure.name)(figure.x, figure.y, self.left))
                self.left_force = self.left_force + 1
                self.won_prize += int(figure.figure.price / 10)
            if self.left_force <= 0:
                self.winner = ""
            return

        if self.right is None:
            self.right = data['user']
            for figure in figures:                
                self.game_field.setFigure(Figures.get_figure(name=figure.figure.name)(Constants.x_block_amount - figure.x - 1, figure.y, self.right))
                self.right_force = self.right_force + 1
                self.won_prize += int(figure.figure.price / 10)
            if self.right_force <= 0:
                self.winner = ""
            else:
                self.start()
            return

    def handleClientMessage(self, data):
        self.turn = abs(self.turn - 1)
        data['turn'] = self.left if self.turn == 0 else self.right
        for obj in data['actions']:
            from_x = obj['from_x']
            to_x = obj['to_x']
            if obj['master'] == self.right:
                obj['from_x'] = Constants.x_block_amount - obj['from_x'] - 1
                obj['to_x'] = Constants.x_block_amount - obj['to_x'] - 1
            if obj['type'] == 'attack':
                self.game_field.attackFigure(obj['from_x'], obj['from_y'], obj['to_x'], obj['to_y'])
                if isinstance(self.game_field.field_matrix[obj['to_x']][obj['to_y']], Figure):
                    if self.game_field.field_matrix[obj['to_x']][obj['to_y']].isDead():
                        if self.turn == 0:
                            self.left_force = self.left_force - 1
                        elif self.turn == 1:
                            self.right_force = self.right_force - 1
                        self.game_field.field_matrix[obj['to_x']][obj['to_y']] = GameObject(obj['to_x'], obj['to_y'])
            if obj['type'] == 'move':
                self.game_field.moveFigure(obj['from_x'], obj['from_y'], obj['to_x'], obj['to_y'])
            obj['from_x'] = from_x
            obj['to_x'] = to_x
        if self.left_force == 0 and self.right_force == 0:
            self.winner = None
        elif self.left_force == 0:
            self.winner = self.right
        elif self.right_force == 0:
            self.winner = self.left

    def start(self):
        self.callback({
            'type': 'init',
            'turn': self.left if self.turn == 0 else self.right,
            'field': self.game_field.getReport(),
            'left': self.left,
            'right': self.right
        })
